coqpyt/coq/proof_file.py

```diff
@@ -796,6 +804,79 @@ def exec(self, nsteps=1) -> List[Step]:
         last, slice = sign == 1, (initial_steps_taken, self.steps_taken)
         return self.steps[slice[1 - last] : slice[last]]
 
+    def next_unproven_proof(self) -> Optional[ProofTerm]:
+        """Gets the next unproven proof.
+
+        Returns:
+            Optional[ProofTerm]: The next unproven proof or None if there are no
+                unproven proofs.
+        """
+        for proof in self.proofs:
+            if not self.__is_proven(proof):
+                return proof
+
+        if len(self.open_proofs) > 0:
+            return self.open_proofs[0]
+
+        return None
+
+    def append_step(self, proof: ProofTerm, step_text: str):
+        """Appends a step to a proof. This function will change the original file.
+        If an exception is thrown the file will not be changed.
+
+        Args:
+            proof (ProofTerm): The proof of the file to which the step is added.
+            step_text (str): The text of the step to add.
+
+        Raises:
+            InvalidFileException: If the file being changed is not valid.
+            InvalidAddException: If the file is invalid after adding the step.
+        """
+        step_index = self.__find_step_index(proof.ast.range)
+        self.add_step(step_index + len(proof.steps), step_text)
+        if self.steps_taken == step_index + len(proof.steps) + 1:
+            self.exec(1)
+
+    def pop_step(self, proof: ProofTerm):
+        """Removes the last step from a proof. This function will change the original file.
+        If an exception is thrown the file will not be changed.
+
+        Args:
+            proof (ProofTerm): The proof of the file from which the step is removed.
+
+        Raises:
+            InvalidFileException: If the file being changed is not valid.
+            InvalidAddException: If the file is invalid after removing the step.
+        """
+        step_index = self.__find_step_index(proof.ast.range)
```

Comment:
More of a comment than a suggestion: raising `ValueError` doesn't seem very informative for the user, since they do not provide the range explicitly. An error message similar to "pop from empty list" would make more sense.